Fix [src] hyperlinks in rustdoc

ipetkov opened the pull request:
* rustdoc was doubly appending the file name to the path of where to
  generate the source files, meanwhile, the [src] hyperlinks were not
* Added a flag to rustdoc::html::render::clean_srcpath to ignore the
  last path component, i.e. the file name itself to prevent the issue
* This also avoids creating directories with the same name as source
  files, and it makes sure the link to `main.css` is correct as well.

Fixes #23192

rust-highfive commented:
r? @huonw

(rust_highfive has picked a reviewer for you, use r? to override)

Gankro commented on src/librustdoc/html/render.rs:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2013-2014 The Rust Project Developers. See the COPYRIGHT
+// Copyright 2013-2015 The Rust Project Developers. See the COPYRIGHT
```
We've basically agreed to stop touching these. They're legal cargo-culting.

frewsxcv replied:
@Gankro https://github.com/rust-lang/rust/issues/9070#issuecomment-72514411

Gankro replied:
Fair enough.

Gankro commented:
Wouls it be possible to write a regression test for this?

steveklabnik commented:
:confetti_ball:

tomjakubowski commented:
@ipetkov if you're not familiar with rustdoc integration/regression testing, look in the `src/test/run-make/rustdoc-*` directories. They are run with a Python script located at `etc/htmldocck.py`; in brief parses directives from comments and tries to match text/patterns at a given XPath.

Gankro commented:
@tomjakubowski Are you interested in being the unofficial-official reviewer for this? You're the most active in this codebase recently.

ipetkov commented:
@tomjakubowski I wasn't until today, but it seemed pretty straight forward :)

Also the PR has been updated to include a regression test

tomjakubowski commented:
@ipetkov looks good to me, but I'd prefer if this regression had its own test. `rustdoc-smoke` is already vague enough and I'd rather it not get cluttered over time.

I'm also a little bit curious why this bug only cropped up recently. Generating `[src]` links worked fine  (minus some special cases like the `alloc` crate bug) until a week or so ago and then it broke seemingly out of nowhere.

ipetkov commented:
@tomjakubowski I'm really not sure what caused the breakage, though I have a feeling it might be related with reforming `PathBuf` because in the past it might not have included the file name when it was converted into an iterator (though I have not looked into it and I could be wrong).

@Gankro @tomjakubowski I've updated the PR so that the regression tests are in their own test rather than in `rustdoc-smoke` and rebased against the updated master.

Sorry for the delay at updating this, my machine suffered from a hardware failure a few nights ago and I haven't had access to my files. I haven't tested (or built) these new updates, though I tried to double check the test cases, and hopefully the build will pass.

huonw commented on src/librustdoc/html/render.rs:
```diff
@@ -695,13 +695,18 @@ fn shortty(item: &clean::Item) -> ItemType {
 // FIXME (#9639): The closure should deal with &[u8] instead of &str
 // FIXME (#9639): This is too conservative, rejecting non-UTF-8 paths
-fn clean_srcpath<F>(src_root: &Path, p: &Path, mut f: F) where
+fn clean_srcpath<F>(src_root: &Path, p: &Path, keep_filename: bool, mut f: F) where
```
Could you add a brief mention of what `keep_filename` does to the docs?

ipetkov commented:
@huonw Updated with docs!

huonw commented:
@bors r+

bors commented:
:pushpin: Commit cc27f29 has been approved by `huonw`

<!-- @bors r=huonw cc27f294da7145022747aefffba3a04c2b2de539 -->

bors commented:
:hourglass: Testing commit cc27f29 with merge 8165036...

Manishearth commented:
Test breaks on a mac?

http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/4122/steps/test/logs/stdio

```
----- /Users/rustbuild/src/rust-buildbot/slave/auto-mac-64-nopt-t/build/src/test/run-make/rustdoc-src-links/ --------------------
------ stdout ---------------------------------------------
DYLD_LIBRARY_PATH="/Users/rustbuild/src/rust-buildbot/slave/auto-mac-64-nopt-t/build/obj/x86_64-apple-darwin/test/run-make/rustdoc-src-links:/Users/rustbuild/src/rust-buildbot/slave/auto-mac-64-nopt-t/build/obj/x86_64-apple-darwin/stage2/lib:" /Users/rustbuild/src/rust-buildbot/slave/auto-mac-64-nopt-t/build/obj/x86_64-apple-darwin/stage2/bin/rustdoc -w html -o /Users/rustbuild/src/rust-buildbot/slave/auto-mac-64-nopt-t/build/obj/x86_64-apple-darwin/test/run-make/rustdoc-src-links/doc foo.rs
/System/Library/Frameworks/Python.framework/Versions/2.7/Resources/Python.app/Contents/MacOS/Python /Users/rustbuild/src/rust-buildbot/slave/auto-mac-64-nopt-t/build/src/etc/htmldocck.py /Users/rustbuild/src/rust-buildbot/slave/auto-mac-64-nopt-t/build/obj/x86_64-apple-darwin/test/run-make/rustdoc-src-links/doc foo.rs

------ stderr ---------------------------------------------
Traceback (most recent call last):
  File "/Users/rustbuild/src/rust-buildbot/slave/auto-mac-64-nopt-t/build/src/etc/htmldocck.py", line 396, in <module>
    check(sys.argv[1], get_commands(sys.argv[2]))
  File "/Users/rustbuild/src/rust-buildbot/slave/auto-mac-64-nopt-t/build/src/etc/htmldocck.py", line 389, in check
    c.cmd, c.lineno))
RuntimeError: @has check failed at line 42
make[1]: *** [all] Error 1
```

bors commented:
:snowman: The build was interrupted to prioritize another pull request.

bors commented:
:hourglass: Testing commit cc27f29 with merge a30341f...

bors commented:
:snowman: The build was interrupted to prioritize another pull request.

bors commented:
:hourglass: Testing commit cc27f29 with merge 10d5cdf...

ipetkov commented:
@huonw @Manishearth d'oh copy paste error in a test file. I was able to test a fix locally and the update should fix things now!

Manishearth commented:
@bors: r=huonw

bors commented:
:pushpin: Commit af6cf85 has been approved by `huonw`

<!-- @bors r=huonw af6cf85b9815644c77c77cdd141d476efc426d99 -->

bors commented:
:hourglass: Testing commit af6cf85 with merge 81e2396...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/4127), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/4131), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/4113), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/4121), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/4094), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/4148), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/4147), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/4134), [auto-win-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-32-nopt-t/builds/3807), [auto-win-32-opt](http://buildbot.rust-lang.org/builders/auto-win-32-opt/builds/3794), [auto-win-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-64-nopt-t/builds/2624), [auto-win-64-opt](http://buildbot.rust-lang.org/builders/auto-win-64-opt/builds/2631)
